File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.js
function solution(board, moves) {
    const N = board.length
    let topIdx = Array(N).fill(-1);
    for(let i = 0 ; i < N ; i++){
        for(let j = 0 ; j < N ; j++){
            if(topIdx[j] === -1 && board[i][j] !== 0) topIdx[j] = i;
        }
    }
    
    let result = 0;
    let stack = [];
    moves.forEach((move) => {
        const curIdx = topIdx[move-1];
        if(curIdx !== -1 && curIdx < N){
            if(stack[stack.length-1] === board[curIdx][move-1]){
                result += 2;
                stack.pop();
            } else {
                stack.push(board[curIdx][move-1]);
            }
            topIdx[move-1] = curIdx + 1;
        }
    })
    return(result);
}

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.py
# 최상단의 위치를 찾을 필요가 없고 0이 아닐때까지 반복문을 돌면 된다 오메
def solution(board, moves):
    N = len(board)
    stack = []
    result = 0
    for move in moves:
        for j in range(N):
            curVal = board[j][move-1]
            if curVal != 0:
                if stack and stack[-1] == curVal:
                    result += 2
                    stack.pop()
                else:
                    stack.append(curVal)
                board[j][move-1] = 0
                break
    return result